src/storage/src/source/postgres.rs

```diff
@@ -129,8 +129,44 @@ impl SourceRender for PostgresSourceConnection {
         Stream<G, (usize, HealthStatusUpdate)>,
         Rc<dyn Any>,
     ) {
-        let resume_upper =
-            Antichain::from_iter(config.source_resume_upper.iter().map(MzOffset::decode_row));
+        // Get the uppers of all exports.
+        let mut export_resume_uppers: BTreeMap<_, _> = config
+            .source_resume_upper
+            .iter()
+            .map(|(id, upper)| {
+                (
+                    *id,
+                    Antichain::from_iter(upper.iter().map(MzOffset::decode_row)),
+                )
+            })
+            .collect();
+
+        // The primary source's upper is not meaningful shard, so ensure it's removed from
+        // consideration.
+        //
+        // TODO: close this upper because it can never do anything.
+        export_resume_uppers.remove(&config.id);
+
+        // The resume_upper is the point at which we will resume replication.
+        let mut resume_upper = Antichain::new();
+        for f in export_resume_uppers
+            .values()
+            // In the future, we will add subsources to export_resume_uppers with the minimum frontier,
+            // which will signal that it is a new collection that needs to be snapshot; this means we do
+            // not want to factor those values in to determining the resume upper.
+            .filter(|u| u.elements() != [MzOffset::minimum()])
```

Comment:
This doesn't look right (and the if statement below is even stranger). Until we have per-subsource handling we must retain the old behavior here, i.e 

```rust
let resume_upper = Antichain::from_iter(export_resume_uppers.values().flat_map(|f| f.elements()).cloned());
```

Reply:
I rewrote this to be in terms of collections rather than their frontier's elements. I do want to keep the behavior of separating collections we need to snapshot from those we don't, though, as a vouch that the approach for adding subsources generalizes.